src/tools/miri/tests/fail/never_say_never.rs

```diff
@@ -6,10 +6,8 @@
 
 fn main() {
     let y = &5;
-    let x: ! = unsafe {
-        *(y as *const _ as *const !) //~ ERROR: entering unreachable code
-    };
-    f(x)
+    let x: ! = unsafe { *(y as *const _ as *const !) };
+    f(x) //~ ERROR: entering unreachable code
 }
```

Comment:
This is surprising. How does the MIR change here? The previous MIR didn't even contain an `f` call, even with `-Zmir-opt-level=0`

Reply:
The change is essentially
1. `f` now has `return`, not `unreachable`
2. `main` now has additional `_3: !`
3. `main` now has `_3 = f(const ZeroSized: !);`, not `unreachable;`

Reply:
Hm, it looks like the difference is already present after building MIR:

```diff
--- ./nightly_mir_dump/t.f.001-000.built.after.mir      2023-05-15 14:25:01.537195340 +0000
+++ ./b_mir_dump/t.f.001-000.built.after.mir    2023-05-15 14:25:07.469143513 +0000
@@ -3,23 +3,9 @@
 fn f(_1: !) -> ! {
     debug x => _1;                       // in scope 0 at ./t.rs:9:6: 9:7
     let mut _0: !;                       // return place in scope 0 at ./t.rs:9:15: 9:16
-    let mut _2: !;                       // in scope 0 at ./t.rs:9:17: 11:2
-    let mut _3: !;                       // in scope 0 at ./t.rs:10:5: 10:6
 
     bb0: {
-        StorageLive(_2);                 // scope 0 at ./t.rs:9:17: 11:2
-        StorageLive(_3);                 // scope 0 at ./t.rs:10:5: 10:6
-        _3 = _1;                         // scope 0 at ./t.rs:10:5: 10:6
-        unreachable;                     // scope 0 at ./t.rs:10:5: 10:6
-    }
-
-    bb1: {
-        StorageDead(_3);                 // scope 0 at ./t.rs:10:5: 10:6
-        unreachable;                     // scope 0 at ./t.rs:9:17: 11:2
-    }
-
-    bb2: {
-        StorageDead(_2);                 // scope 0 at ./t.rs:11:1: 11:2
+        _0 = _1;                         // scope 0 at ./t.rs:10:5: 10:6
         return;                          // scope 0 at ./t.rs:11:2: 11:2
     }
 }
```

Reply:
I don't understand why `NeverToAny` should get a special case for `! -> !`, but I also don't mind the change in diagnostics here.

Reply:
That's somewhat surprising but well, I am not familiar with those parts of the compiler at all.^^
Still, @rust-lang/types should probably take a look.